### PR TITLE
perf: Neo4j csv publisher using apoc library for performance improvements - CLEAN

### PR DESCRIPTION
**Pull request by zacr**
### Summary of Changes

New Neo4JCsvPublisher implementation using the APOC library for 5x performance improvement. This is a fresh PR because [original one](https://github.com/amundsen-io/amundsen/pull/1836) wasn't created on a clean branch.

### Tests

I didn't see any automated tests for Neo4JCsvPublisher. This PR was tested manually by several community members.

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [ ] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [ ] PR includes a summary of changes.
- [ ] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does

### CODE REVIEW

**Review by akiratwang — COMMENTED** [review]
Looks great! Once this is merged, I'll deploy into our dev environment and start testing other features out.

AFAIK this doesn't support multiple programmatic descriptions being added, so I'll work on that PR post-merge. 

Excellent stuff!

**Comment by chonyy** [comment]
Looking forward to having this PR merged! Would like to use this in the official build 🚀

**Comment by akiratwang** [comment]
Ahh looks like `DCO` is blocked. Can you randomly add a small commit @zacr with format:

`Signed-off-by: your_name_here <your_email_here>`

(https://github.com/amundsen-io/amundsen/pull/1877/checks?check_run_id=6632089423)

**Comment by akiratwang** [comment]
> 

Ahh yep, that would make sense with the `.lower()` I saw in the previous iteration. Thanks for investigating this as I'm not too familiar with Neo4J!

**Comment by zacr** [comment]
Thanks so much @chonyy. I have an update in progress that addresses all three, hope to have it committed in the next day or so.

**Comment by chonyy** [comment]
> Thanks so much @chonyy. I have an update in progress that addresses all three, hope to have it committed in the next day or so.

@zacr that's great news! I could also help with testing it in my environment once you are done 🚀

**Comment by chonyy** [comment]
Hey @zacr ,

FYI, I tested your code in neo4j `3.5.26` and apoc `3.5.0.17`, it also doesn't work. The error message is like below.
image

After I changed all the key to uppercase, it works! Just want to let you know that case sensitivity seems to also be an issue in your provided environment.

**Comment by zacr** [comment]
Awesome @chonyy . Agreed on the upper case. Can we assume the headers of the CSV files will always be upper case?

**Comment by zacr** [comment]
Also, I am out of town this week, but here is the error handling code that goes right after session.run() in _execute_statement()

                ret = [dict(i) for i in result]

                if ret[0]['failedOperations'] > 0:
                    raise RuntimeError(f"Failed to executed statement: {stmt} with {ret[0]['errorMessages']}")

Was thinking of adding a flag to turn on\off this checking.

**Comment by ozandogrultan** [comment]
Any updates on this PR? Would be great to see it merged!

**Comment by chonyy** [comment]
I'm thinking that maybe we don't even have to spend time supporting Neo4j `3.3.0` since it's already EOL. @zacr's works great with some minor fixes related to case sensitivity. 
What I can help here is to add the fix and also the extra part that zacr mentioned, in order to surface the error to the caller side. But we still need a commiter's help to look into this and approve the CI pipeline run 🥲

File: databuilder/databuilder/publisher/neo4j_csv_publisher_apoc.py [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+    def _publish_node(self, node_file: str) -> None:
+        """
+        Iterate over the csv records of a file, each csv record transform to apoc.node.merge statement and will be executed.
+        All nodes should have a unique key, and this method will try to create unique index on the LABEL when it sees
+        first time within a job scope.
+        Example of Cypher query executed by this method:
+        CALL apoc.periodic.iterate(
+            'UNWIND $rows AS row RETURN row', 
+            'CALL apoc.merge.node([row.label], {key:row.key}, row, row) YIELD node RETURN COUNT(*);', 
+            {batchSize: $batch_size, iterateList: True, parallel: False, params: { rows: $batch, tag: $publish_tag }}
```
Comment (conker84) [inline databuilder/databuilder/publisher/neo4j_csv_publisher_apoc.py:259]:
```suggestion
            {batchSize: $batch_size, parallel: False, params: { rows: $batch, tag: $publish_tag }}
```

I would like to remove `iterateList` as it's deprecated and use the default `batchMode` which is `BATCH`, that has the same behaviour.

File: databuilder/databuilder/publisher/neo4j_csv_publisher_apoc.py [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+            return """            
+                    CALL apoc.periodic.iterate('UNWIND $rows AS row RETURN row', '
+                        CALL apoc.merge.node([row.label], {key:row.key}, row, {published_tag:$tag,publisher_last_updated_epoch_ms:timestamp()}) YIELD node RETURN COUNT(*);
+                    ', {batchSize: $batch_size, iterateList: True, parallel: False, params: { rows: $batch, tag: $publish_tag }})
+                """
+
+        return """            
+            CALL apoc.periodic.iterate('UNWIND $rows AS row RETURN row', '
+                CALL apoc.merge.node([row.label], {key:row.key}, row, row) YIELD node RETURN COUNT(*);
+            ', {batchSize: $batch_size, iterateList: True, parallel: False, params: { rows: $batch }})
```
Comment (conker84) [inline databuilder/databuilder/publisher/neo4j_csv_publisher_apoc.py:321]:
```suggestion
            ', {batchSize: $batch_size, parallel: False, params: { rows: $batch }})
```
like above

**Review by conker84 — COMMENTED** [review]
Hi I just checked your code and generally seems good to me, I have just a couple of nitpicks and one related to the index creation that it's a bug (so the performance improvement could be greater than your initial estimation).

One important thing that seems missing is the support for a database that is not the default one. Neo4j since version 4.X supports multi-tenancy so it would be good to have it in there.

**Comment by kristenarmes** [comment]
Hello, I know it has been a while since there has been activity on this PR, but I wanted to follow up for those who were interested in a faster publisher. A few months ago I worked on one myself in [this PR](https://github.com/amundsen-io/amundsen/pull/1957) using the `unwind` clause, which improves performance by allowing Neo4j to compile and cache the statements and reduces the overall amount of transactions ([see it here](https://github.com/amundsen-io/amundsen/blob/main/databuilder/databuilder/publisher/neo4j_csv_unwind_publisher.py)). This publisher doesn't depend on the apoc library. We have been using it and seeing huge improvements in speed compared to the old one. I haven't tested it on all versions of Neo4j, but it may be sufficient for others to try and see if this in progress one using apoc is no longer required?

**Comment by Golodhros** [comment]
Closing as done.

**Pull request closed without merging**